internal/status/cache.go

```diff
@@ -0,0 +1,155 @@
+package status
+
+import (
+	"fmt"
+
+	projcontour "github.com/projectcontour/contour/apis/projectcontour/v1"
+	"github.com/projectcontour/contour/internal/k8s"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type ConditionType string
+
+const ValidCondition ConditionType = "Valid"
+
+// ProxyUpdate holds status updates for a particular HTTPProxy object
+type ProxyUpdate struct {
+	Object     projcontour.HTTPProxy
+	Conditions map[ConditionType]*projcontour.DetailedCondition
+}
+
+// ConditionFor returns a DetailedCondition for a given ConditionType.
+// Currently only "Valid" is supported.
+func (pu *ProxyUpdate) ConditionFor(cond ConditionType) *projcontour.DetailedCondition {
+
+	if cond == "" {
+		return nil
+	}
+
+	dc, ok := pu.Conditions[cond]
+	if !ok {
+		newDc := &projcontour.DetailedCondition{}
+		newDc.Type = string(cond)
+
+		pu.Conditions[cond] = newDc
+		return newDc
+	}
+	return dc
+}
+
+func (pu *ProxyUpdate) StatusMutatorFunc() k8s.StatusMutator {
+
+	return k8s.StatusMutatorFunc(func(obj interface{}) interface{} {
+		switch o := obj.(type) {
+		case *projcontour.HTTPProxy:
+			proxy := o.DeepCopy()
+
+			for condType, cond := range pu.Conditions {
+				condIndex := projcontour.GetConditionIndex(string(condType), proxy.Status.Conditions)
+				if condIndex >= 0 {
+					proxy.Status.Conditions[condIndex] = *cond
+				} else {
+					proxy.Status.Conditions = append(proxy.Status.Conditions, *cond)
+				}
+
+			}
+
+			// Set the old status fields using the Valid DetailedCondition's details.
+			// Other conditions are not relevant for these two fields.
+			validCondIndex := projcontour.GetConditionIndex(string(ValidCondition), proxy.Status.Conditions)
+			if validCondIndex >= 0 {
+				validCond := proxy.Status.Conditions[validCondIndex]
+				switch validCond.Status {
+				case projcontour.ConditionTrue:
+					orphanCond, orphaned := validCond.GetError(k8s.StatusOrphaned)
+					if orphaned {
+						proxy.Status.CurrentStatus = k8s.StatusOrphaned
+						proxy.Status.CurrentStatus = orphanCond.Message
+					}
+					proxy.Status.CurrentStatus = k8s.StatusValid
+					proxy.Status.Description = validCond.Reason + ": " + validCond.Message
+				case projcontour.ConditionFalse:
+					proxy.Status.CurrentStatus = k8s.StatusInvalid
+					proxy.Status.Description = validCond.Reason + ": " + validCond.Message
+				}
+			}
+
+			return proxy
+		default:
+			panic(fmt.Sprintf("Unsupported object %s/%s in status Address mutator",
```

Comment:
This package controls the mutator, and it's should only ever be passed HTTPProxy objects. If we get to this code, something has gone seriously wrong. That's why the panic.

Reply:
Seems lame to panic and break someone's application. Should a non-HTTPProxy object get here, log it and move on. There are other places that we panic as well for similar conditions. To me feels like ignoring that case is fine since it would have no affect anyway.